compiler/planner/insert_planner.rs

```diff
@@ -0,0 +1,402 @@
+    extend_type_constants_for_untyped_role_players_from_annotations(
+        constraints,
+        type_annotations,
+        &mut type_constants,
+    )?;
+    let value_constants = collect_values_from_expression_and_comparison(constraints, &input_variables)?;
+    let isa_kinds = collect_kind_from_isa(constraints, input_variables, type_annotations)?;
+
+    let mut instructions = Vec::new();
+    filter_variants!(Constraint::Isa : constraints).enumerate().try_for_each(|(i, isa)| {
+        debug_assert!(*isa_kinds.index.get(&isa.thing()).unwrap() == i);
+        let type_ = get_type_source(&input_variables, &type_constants.index, isa.type_())?;
+        let kind = isa_kinds.items[i];
+        let instruction = match kind {
+            Kind::Entity => InsertInstruction::Entity { type_ },
+            Kind::Attribute => {
+                let value = get_value_source(&input_variables, &value_constants.index, isa.thing())?;
+                InsertInstruction::Attribute { type_, value }
+            }
+            Kind::Relation => InsertInstruction::Relation { type_ },
+            Kind::Role => Err(InsertCompilationError::IsaStatementForRoleType { isa: isa.clone() })?,
+        };
+        instructions.push(instruction);
+        Ok(())
+    })?;
+
+    for constraint in constraints {
+        match constraint {
+            Constraint::Has(has) => {
+                instructions.push(InsertInstruction::Has {
+                    owner: get_thing_source(input_variables, &isa_kinds.index, has.owner())?,
+                    attribute: get_thing_source(input_variables, &isa_kinds.index, has.attribute())?,
+                });
+            }
+            Constraint::RolePlayer(role_player) => {
+                instructions.push(InsertInstruction::RolePlayer {
+                    relation: get_thing_source(input_variables, &isa_kinds.index, role_player.relation())?,
+                    player: get_thing_source(input_variables, &isa_kinds.index, role_player.player())?,
+                    role: get_type_source(input_variables, &type_constants.index, role_player.role_type())?,
+                });
+            }
+            Constraint::Isa(_)
+            | Constraint::Label(_)
+            | Constraint::RoleName(_)
+            | Constraint::ExpressionBinding(_)
+            | Constraint::Comparison(_) => {} // already handled
+            Constraint::Sub(_) | Constraint::FunctionCallBinding(_) => unreachable!(),
+        }
+    }
+
+    Ok(assemble_plan(instructions, value_constants, type_constants, isa_kinds))
+}
+
+fn assemble_plan(
+    instructions: Vec<InsertInstruction>,
+    value_sources: Sources<Value<'static>>,
+    type_sources: Sources<Type>,
+    kind_sources: Sources<Kind>,
+) -> InsertPlan {
+    // TODO: Maybe take output row as parameter?
+    let output_row = kind_sources
+        .index
+        .iter()
+        .map(|(_, i)| VariableSource::ThingSource(ThingSource::Inserted(*i)))
+        .collect::<Vec<_>>();
```

Comment:
Should the output row be planned? Or should we just return `concat(copy(input_row), inserted)`